### ISSUE TITLE
Touch() does not work

### ISSUE THREAD

**Issue by moyarich** [phpseclib/phpseclib#1052]
The sftp->touch() function does not work. It creates a 0 filesize file on the server instead of modifying the mtime and atime of the original file

**Comment by terrafrost** [comment]
I tried it out and it works as intended. The fact that you didn't say that `touch()` deleted a file suggests to me that the file you were "touching" didn't exist and modifying the mtime and atime of a non-existent file is quite impossible.

Quoting the phpDoc comments,

```php
    /**
     * Sets access and modification time of file.
     *
     * If the file does not exist, it will be created.
     *
     * @param string $filename
     * @param int $time
     * @param int $atime
     * @throws \UnexpectedValueException on receipt of unexpected packets
     * @return bool
     * @access public
     */
```

This behavior is inspired by PHP's own touch function. Quoting http://php.net/manual/en/function.touch.php ,

> Attempts to set the access and modification times of the file named in the **filename** parameter to the value given in **time**. Note that the access time is always modified, regardless of the number of parameters.
> 
> If the file does not exist, it will be created.

**Comment by moyarich** [comment]
The file existed on the server, before I used the touch function
 here a snippet of the code I used.
```php

$sourceFilePath = getcwd() . DIRECTORY_SEPARATOR .'test_file1.txt';
$destFilePath = 'test_file1_ftp.txt';


//** copy file to ftp server
$copyResult = $sftp->put($destFilePath, $sourceFilePath, $sftp::SOURCE_LOCAL_FILE);

$sourceModifiedTime = filemtime($sourceFilePath);
$sourceAccessTime   = fileatime($sourceFilePath);
$sftp->touch($destFilePath,$sourceModifiedTime,$sourceAccessTime);

**Comment by bantu** [comment]
@moyarich Are you saying that touch() truncated the file down to 0 bytes?

**Comment by terrafrost** [comment]
It's working fine for me:

```php
$ssh->put('test.ext', 'test.ext', NET_SFTP_LOCAL_FILE);

$sourceModifiedTime = filemtime('test.ext');
$sourceAccessTime   = fileatime('test.ext');

echo 'modified time: ' . date('Y-m-d', $sourceModifiedTime) . "\n";
echo 'access time:   ' . date('Y-m-d', $sourceAccessTime) . "\n";
$ssh->touch('test.ext',$sourceModifiedTime,$sourceAccessTime);
```

I then do stat test.ext on the server that I uploaded the file to and ghe access and modify dates are as expected..

As bantu's post points out it's not exactly clear what the problem you're having is but whatever it is the SFTP logs might provide insight.

**Comment by moyarich** [comment]
The file is 0 bytes

**Comment by bantu** [comment]
Please describe your environment. Which SFTP server is this?

**Comment by moyarich** [comment]
are you asking what version of phpseclib i am using?
I used composer to install it
"phpseclib/phpseclib": "^2.0"

there is no version # in the SFT.php file, but i do see this in the readme file:
[![Build Status](https://travis-ci.org/phpseclib/phpseclib.svg?branch=2.0)]

**Comment by bantu** [comment]
@moyarich No, I am asking about the SSH/SFTP server you're connecting to. Can you share the hostname/IP?

**Comment by moyarich** [comment]
ooh sorry, it is a keyfile protected sftp server, I would have to share the info privately

**Comment by moyarich** [comment]
Id ont know how usefult this information is but when i do $sftp->getServerIdentification()

i get the result : **SSH-2.0-ssh2js0.1.11srv**

**Comment by terrafrost** [comment]
Is https://github.com/mscdex/ssh2 what you're using as the server? Try doing these actions with some client other than phpseclib and see if the same thing doesn't happen.

**Comment by terrafrost** [comment]
So I installed https://github.com/mscdex/ssh2 locally and that does have the same identification string that you're reporting. But that library just stubs things out on the server side. It has a [bunch of events](https://github.com/mscdex/ssh2-streams/blob/master/SFTPStream.md) that you can attach handlers too but it's up to you to actually write the handlers to interface with the operating system.

`is_dir` would trigger the STAT event and `touch` would trigger the SETSTAT event. If the file didn't exist `OPEN` might additionally be called.

So, basically, this is sounding like it's not a bug in mscdex/ssh2 but in the event handlers that I assume you wrote.
